include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+                        arma::kron(Vector(M_, arma::fill::ones), umax).t());
+    size_t Aineq_dim = N_ * n_ + M_ * m_;
+    Aineq_ = arma::eye<arma::SpMat<data_t>>(Aineq_dim, Aineq_dim);
+  }
+
+  void Print() {
+    sys_.Print();
+    // TODO: Implement print
+  }
+
+ protected:
+  System sys_;  ///< system being controlled
+  size_t n_;    ///< number of states
+  size_t m_;    ///< number of output states
+  size_t N_;    ///< number of steps
+  size_t M_;    ///< number of inputs
+  Matrix A_;    ///< state transition matrix
+  Matrix B_;    ///< input matrix
+  Matrix P_;    ///< penalty matrix
+  Matrix Q_;
+  Matrix S_;
+
+  Matrix lineq_;               ///< lower inequality bound
+  Matrix uineq_;               ///< upper inequality bound
+  arma::SpMat<data_t> Aineq_;  ///< inequality condition matrix
+
+  Matrix Acon_;  ///< update condition matrix
+  Vector lb_;    ///< lower bound
+  Vector ub_;    ///< upper bound
+
+  Vector xi_;     ///< previous step end state
+  size_t t_sim_;  ///< previous step simulation time step
+
+ private:
+  /**
+   * @brief     Set the matrix for the OSQP solver from an Armadillo matrix
+   *
+   * @param     A The matrix to convert
+   */
+  OSQPCscMatrix* from_matrix(const Matrix& A);
+
+  /**
+   * @brief     Set the matrix for the OSQP solver from an Armadillo sparse
+   * matrix
+   *
+   * @param     A The sparse matrix to convert
+   */
+  OSQPCscMatrix* from_sparse(const arma::SpMat<data_t>& A);
+
+  /**
+   * @brief     Set the vector for the OSQP solver from an Armadillo vector
+   *
+   * @param     v The vector to convert
+   */
+  OSQPFloat* from_vec(const Vector& v);
+
+  /**
+   * @brief     Calculate the trajectory for the simulation step,
+   *            used when the simulation time step is the same as
+   *            the prior step
+   *
+   * @param     x0  The initial state
+   * @param     u0  The initial control input
+   * @param     xr  The reference trajectory
+   * @param     out Print out step information
+   *
+   * @return    The trajectory for the simulation step
+   */
+  OSQPSolution* fast_update(const Vector& x0, const Vector& u0,
+                            const Matrix& xr, size_t n_sim);
+
+  /**
+   * @brief     Calculate the trajectory for the simulation step
+   *
+   * @param     x0  The initial state
+   * @param     u0  The initial control input
+   * @param     xr  The reference trajectory
+   * @param     out Print out step information
+   *
+   * @return    The trajectory for the simulation step
+   */
+  OSQPSolution* slow_update(const Vector& x0, const Vector& u0,
+                            const Matrix& xr, size_t n_sim);
+
+  /**
+   * @brief     Create an identity matrix with an offset axis
+   * 
+   * @param     n The size of the matrix
+   * @param     k The offset axis
+   * 
+   * @return    The identity matrix with an offset axis
+   */
+  arma::SpMat<data_t> eye_offset(size_t n, int k = -1) {
```

Comment:
Not sure if it helps if its output is dense, but FYI arma has a function to do this: https://arma.sourceforge.net/docs.html#diagmat

Reply:
I didn't see that! I think that will make it easier to do.